File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/MinimalistCollectionViewer.java
package joookiwi.collection.java;

import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static java.lang.Integer.MAX_VALUE;

/// A [MinimalistCollectionHolder] that hold another [MinimalistCollectionHolder]
/// (which could be [mutable][MutableMinimalistCollectionHolder])
/// to only view it
///
/// @implNote It does not change any behaviour from its [reference][#getReference]
/// @see CollectionViewer
@NotNullByDefault
public class MinimalistCollectionViewer<T extends @Nullable Object,
        REFERENCE extends MinimalistCollectionHolder<T>>
        extends AbstractMinimalistCollectionHolder<T> {

    private final REFERENCE __reference;

    public MinimalistCollectionViewer(final REFERENCE reference) {
        super();
        this.__reference = reference;
    }

    public REFERENCE getReference() { return __reference; }

    @Override public @Range(from = 0, to = MAX_VALUE) int size() { return getReference().size(); }

    @Override public T get(final int index) { return getReference().get(index); }

}

File: javascript/src/MinimalistCollectionViewer.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {MinimalistCollectionHolder} from "./MinimalistCollectionHolder"

import {AbstractMinimalistCollectionHolder} from "./AbstractMinimalistCollectionHolder"

/**
 * A {@link MinimalistCollectionHolder} that hold another {@link MinimalistCollectionHolder}
 * (which could be mutable)
 * to only view it
 *
 * @apiNote It does not change any behaviour from its {@link reference}
 * @typeParam T         The element type
 * @typeParam REFERENCE The reference passed in the constructor
 */
export class MinimalistCollectionViewer<const T = unknown,
    const REFERENCE extends MinimalistCollectionHolder<T> = MinimalistCollectionHolder<T>, >
    extends AbstractMinimalistCollectionHolder<T> {

    readonly #reference: REFERENCE

    public constructor(reference: REFERENCE,) {
        super()
        this.#reference = reference
    }

    /** The reference to view */
    public get reference(): REFERENCE {
        return this.#reference
    }

    public override get size(): REFERENCE["size"] {
        return this.reference.size
    }

    public override get(index: number,): T {
        return this.reference.get(index,)
    }

}
